gossip/tips_cache.c

```diff
@@ -0,0 +1,218 @@
+  }
+
+  rw_lock_handle_wrlock(&cache->lock);
+  ret = tips_cache_fifo_add(&cache->tips, cache->capacity, tip);
+  rw_lock_handle_unlock(&cache->lock);
+
+  return ret;
+}
+
+retcode_t tips_cache_remove(tips_cache_t* const cache,
+                            flex_trit_t const* const tip) {
+  retcode_t ret = RC_OK;
+
+  if (cache == NULL || tip == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_wrlock(&cache->lock);
+  ret = hash243_set_remove(&cache->tips, tip);
+  rw_lock_handle_unlock(&cache->lock);
+
+  return ret;
+}
+
+retcode_t tips_cache_set_solid(tips_cache_t* const cache,
+                               flex_trit_t const* const tip) {
+  retcode_t ret = RC_OK;
+
+  if (cache == NULL || tip == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_wrlock(&cache->lock);
+  if (hash243_set_contains(&cache->tips, tip)) {
+    if ((ret = hash243_set_remove(&cache->tips, tip)) != RC_OK) {
+      goto done;
+    }
+    if ((ret = tips_cache_fifo_add(&cache->solid_tips, cache->capacity, tip)) !=
+        RC_OK) {
+      goto done;
+    }
+  }
+
+done:
+  rw_lock_handle_unlock(&cache->lock);
+  return ret;
+}
+
+size_t tips_cache_non_solid_size(tips_cache_t* const cache) {
+  size_t size = 0;
+
+  if (cache == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_rdlock(&cache->lock);
+  size = hash243_set_size(&cache->tips);
+  rw_lock_handle_unlock(&cache->lock);
+
+  return size;
+}
+
+size_t tips_cache_solid_size(tips_cache_t* const cache) {
+  size_t size = 0;
+
+  if (cache == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_rdlock(&cache->lock);
+  size = hash243_set_size(&cache->solid_tips);
+  rw_lock_handle_unlock(&cache->lock);
+
+  return size;
+}
+
+size_t tips_cache_size(tips_cache_t* const cache) {
+  size_t size = 0;
+
+  if (cache == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_rdlock(&cache->lock);
+  size = hash243_set_size(&cache->tips) + hash243_set_size(&cache->solid_tips);
+  rw_lock_handle_unlock(&cache->lock);
+
+  return size;
+}
+
+retcode_t tips_cache_random_tip(tips_cache_t* const cache,
+                                flex_trit_t* const tip) {
+  retcode_t ret = RC_OK;
+
+  if (cache == NULL || tip == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_rdlock(&cache->lock);
+
+  if (hash243_set_size(&cache->tips) == 0) {
+    memset(tip, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
+    goto done;
+  }
+  ret = tips_cache_random_tip_from_set(&cache->tips, tip);
+
+done:
+  rw_lock_handle_unlock(&cache->lock);
+  return ret;
+}
+
+retcode_t tips_cache_random_solid_tip(tips_cache_t* const cache,
+                                      flex_trit_t* const tip) {
+  retcode_t ret = RC_OK;
+
+  if (cache == NULL || tip == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  rw_lock_handle_rdlock(&cache->lock);
+
+  if (hash243_set_size(&cache->solid_tips) == 0) {
+    rw_lock_handle_unlock(&cache->lock);
+    return tips_cache_random_tip(cache, tip);
```

Comment:
Calling this method here is counterintuitive? This also relocks the rdlock again

Reply:
Would you return null hash as well if there is no solid hash ?